### PR TITLE
Feature: Crystal Hollows Lobby Age Warning

### PR DESCRIPTION
**Pull request by RayDeeUx**
## Dependencies
- hannibal002/SkyHanni-REPO#93

## Changelog New Features
+ Crystal Hollows Lobby Age Warning. - Erymanthus

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+    }
+
+    private fun processLobbyAge() {
+        if (!isInCrystalHollows()) return //avoid incredibly fringe race condition that somehow only happens in the dwarven mines -ery
+        val lobbyAge = getLobbyAgeInMinecraftDays()
+        if (oldLobbyAge == lobbyAge) return //avoid sending chat msgs more than necessary
+        oldLobbyAge = lobbyAge //set new cache value
+        if (config.lobbyAgeReminders && !lobbyIsPastAgeThreshold) LorenzUtils.chat("§aThis Crystal Hollows lobby is currently at Day $lobbyAge, ${abs(config.minLobbyAgeThreshold - lobbyAge)} day(s) away from Day ${config.minLobbyAgeThreshold}.")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]: could use `StringUtils.optionalPlural` here

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    @Expose
+    @ConfigOption(name = "Enabled", desc = "Toggles the overall warning.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean enabled = false;
+
+    @Expose
+    @ConfigOption(name = "Lobby Age Reminders", desc = "Toggles lobby age reminders before it passes a certain point.")
+    @ConfigEditorBoolean
+    @FeatureToggle
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java]: this doesnt need to be a feature toggle

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public boolean lobbyAgeReminders = false;
+
+    @Expose
+    @ConfigOption(name = "Player Count Reminders", desc = "Toggles player count reminders before it passes a certain point.")
+    @ConfigEditorBoolean
+    @FeatureToggle
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java]: same here

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public int minPlayers = 4; //hidden config option that gets added to config.json in case hypixel changes these particular values and a hotfix isn't immediately possible -ery
+
+    @Expose
+    @ConfigOption(name = "Min Lobby Age Threshold (in MC Days)", desc = "")
+    public int minLobbyAgeThreshold = 18; //hidden config option that gets added to config.json in case hypixel changes these particular values and a hotfix isn't immediately possible -ery
+
+    @Expose
+    @ConfigOption(name = "Max Lobby Age Threshold (in MC Days)", desc = "")
+    public int maxLobbyAgeThreshold = 25; //hidden config option that gets added to config.json in case hypixel changes these particular values and a hotfix isn't immediately possible -ery
+}
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java]: this is why we have a repo, with the repo they can wait the up to 24hrs for up to 3 numbers being changed.
These should not be included as hidden config elements like this

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if (!isInCrystalHollows()) return //avoid incredibly fringe race condition that somehow only happens in the dwarven mines -ery
+        for (line in TabListData.getTabList()) {
+            playerCountTabListPattern.matchMatcher(line) {
+                val playerCount = group("playerCount").formatNumber()
+                if (oldPlayerCount == playerCount) return //avoid sending chat msgs more than necessary
+                oldPlayerCount = playerCount //set new cache value
+                if (config.playerCountReminders) LorenzUtils.chat("§a$playerCount players are currently in this Crystal Hollows lobby.")
+                if (!lobbyIsPastAgeThreshold) return
+                if (playerCount in 1..config.minPlayers) LorenzUtils.chat("§cThere are $playerCount players remaining in this Crystal Hollows lobby. §4§lIt will shut down very soon.")
+                else LorenzUtils.chat("There are $playerCount players remaining in this Crystal Hollows lobby. §cIt will shut down when there are fewer than ${config.minPlayers} people left, or when this lobby reaches Day ${config.maxLobbyAgeThreshold} §e(whichever happens first).")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]: i feel the `whichever happens first` is unneeded

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    @SubscribeEvent
+    fun onIslandChange(event: IslandChangeEvent) {
+        if (event.newIsland != THE_CRYSTAL_HOLLOWS) {
+            lobbyIsPastAgeThreshold = false //allow sending again in next CH lobby
+            oldPlayerCount = 0L //reset cache values
+            oldLobbyAge = 0L //reset cache values
+        }
+    }
+
+    private val config get() = SkyHanniMod.feature.mining.crystalHollowsLobbyAgeWarning
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]: i feel most of our code has the config variable at the top of the file not the bottom

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import net.minecraft.client.Minecraft
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.math.abs
+
+class CrystalHollowsLobbyAgeWarning {
+    private val mc get() = Minecraft.getMinecraft()
+
+    private val THE_CRYSTAL_HOLLOWS = IslandType.CRYSTAL_HOLLOWS
+
+    private val TICKS_PER_MC_DAY_MINECRAFT: Long = 24000L
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]: dont think this needs mc and minecraft in the same const name

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    private val crystalHollowsLobbyRepoGroup = RepoPattern.group("crystalhollows.lobbyage")
+    private val playerCountTabListPattern by crystalHollowsLobbyRepoGroup.pattern(("playercount.tablist"), (" *(?:§.)*Players[\\S ]*[\\(\\[\\{](?<playerCount>[\\S ]+)[\\)\\]\\}]"))
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]: not sure if we need the group here as its only 1 pattern
also there seems to be extra sets of brackets in the .pattern
also up to you but if you format these lines more like i have in https://github.com/hannibal002/SkyHanni/pull/871 (naming of group if you still want it and formatting of the pattern line itself) that would be cool

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

**Comment by CalMWolfs** [comment]
also forgot to say lots of somewhat random comments that seem a little unnecessary
was also goign to say that i dont think you can add desciptions for the config categories but ur weird way of storing config values crashes me anyways https://hst.sh/ovifufebad

File: src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java (hunk 4 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/config/features/mining/CrystalHollowsLobbyAgeWarningConfig.java]: done, see hannibal002/SkyHanni-REPO#93

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
still not sure what race condition those comments are reffering to (and also dont really need to know) but you check if in crystal hollows twice. Once when calling isEnabled and once in the other function. Once one of those is removed then this is looking good (also prob dont need the race condition comments)

**Comment by RayDeeUx** [comment]
should mention: the race condition is that it'll (sometimes) send a chat message on player count (and presumably server day, but player count changes more often) when you go to dwarven mines

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+class CrystalHollowsLobbyAgeWarning {
+    private val config get() = SkyHanniMod.feature.mining.crystalHollowsLobbyAgeWarning
+    private val mc get() = Minecraft.getMinecraft()
+
+    private val THE_CRYSTAL_HOLLOWS = IslandType.CRYSTAL_HOLLOWS
+
+    private val TICKS_PER_MC_DAY: Long = 24000L
+
+    private val crystalHollowsLobbyRepoGroup = RepoPattern.group("chlobbyage")
+    private val playerCountTabListPattern by crystalHollowsLobbyRepoGroup.pattern(
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]:
```suggestion
    private val patternGroup = RepoPattern.group("mining.crystalhollows.lobbyage")
    private val playerCountTabListPattern by patternGroup.pattern(
```

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    
+    @SubscribeEvent
+    fun onRepoReload(event: RepositoryReloadEvent) {
+        val data = event.getConstant<CrystalHollowsLobbyAgeWarningJson>("CrystalHollowsLobbyAgeWarning")
+        minPlayers = data.minPlayers
+        minLobbyAgeMCDays = data.minLobbyAgeMCDays
+        maxLobbyAgeMCDays = data.maxLobbyAgeMCDays
+    }
+
+    private fun isEnabled() = config.enabled
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt]:
```suggestion
    private fun isEnabled() = THE_CRYSTAL_HOLLOWS.isInIsland() && config.enabled
```

and then can remove all other checks in the code for being in skyblock and in the crystal hollows

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
a few more code cleanup things

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    private val mc get() = Minecraft.getMinecraft()
+
+    private val THE_CRYSTAL_HOLLOWS = IslandType.CRYSTAL_HOLLOWS
+
+    private val TICKS_PER_MC_DAY: Long = 24000L
+
+    private val patternGroup = RepoPattern.group("mining.crystalhollows.lobbyage")
+    private val playerCountTabListPattern by patternGroup.pattern(
+        "playercount.tablist",
+        " *(?:§.)*Players[\\S ]*[({\\[](?<playerCount>[\\S ]+)[)}\\]]"
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:32]:
```suggestion
        "\\s+(?:§.)+Players\\s+(?:§.)+\\((?<playerCount>\\d+)\\)"
```

this should be better

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+                val playerCount = group("playerCount").formatNumber()
+                if (oldPlayerCount == playerCount) return
+                oldPlayerCount = playerCount
+                if (config.playerCountReminders) LorenzUtils.chat("§a$playerCount players are currently in this Crystal Hollows lobby.")
+                if (!lobbyIsPastAgeThreshold) return
+                if (playerCount in 1..minPlayers) LorenzUtils.chat("§cThere are $playerCount players remaining in this Crystal Hollows lobby. §4§lIt will shut down very soon.")
+                else LorenzUtils.chat("There are $playerCount players remaining in this Crystal Hollows lobby. §cIt will shut down when there are fewer than $minPlayers people left, or when this lobby reaches Day $maxLobbyAgeMCDays.")
+            }
+        }
+    }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:56]: why new function

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    private fun processPlayerCount() {
+        for (line in TabListData.getTabList()) {
+            playerCountTabListPattern.matchMatcher(line) {
+                val playerCount = group("playerCount").formatNumber()
+                if (oldPlayerCount == playerCount) return
+                oldPlayerCount = playerCount
+                if (config.playerCountReminders) LorenzUtils.chat("§a$playerCount players are currently in this Crystal Hollows lobby.")
+                if (!lobbyIsPastAgeThreshold) return
+                if (playerCount in 1..minPlayers) LorenzUtils.chat("§cThere are $playerCount players remaining in this Crystal Hollows lobby. §4§lIt will shut down very soon.")
+                else LorenzUtils.chat("There are $playerCount players remaining in this Crystal Hollows lobby. §cIt will shut down when there are fewer than $minPlayers people left, or when this lobby reaches Day $maxLobbyAgeMCDays.")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:53]:
```suggestion
                else LorenzUtils.chat("There are $playerCount players remaining in this Crystal Hollows lobby. §cIt will shut down when there are fewer than $minPlayers people left, or when this lobby reaches Day $maxLobbyAgeMCDays.")
                return
```

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    )
+
+    private var lobbyIsPastAgeThreshold: Boolean = false
+    private var oldPlayerCount: Long = 0L
+    private var oldLobbyAge: Long = 0L
+    private var minPlayers: Int = 4
+    private var minLobbyAgeMCDays: Int = 18
+    private var maxLobbyAgeMCDays: Int = 25
+
+    private fun getLobbyAgeInMinecraftDays(): Long = (mc.theWorld.worldTime) / TICKS_PER_MC_DAY //world day changes at 6AM in skyblock time
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:42]: This shouldnt be a long

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    private val patternGroup = RepoPattern.group("mining.crystalhollows.lobbyage")
+    private val playerCountTabListPattern by patternGroup.pattern(
+        "playercount.tablist",
+        " *(?:§.)*Players[\\S ]*[({\\[](?<playerCount>[\\S ]+)[)}\\]]"
+    )
+
+    private var lobbyIsPastAgeThreshold: Boolean = false
+    private var oldPlayerCount: Long = 0L
+    private var oldLobbyAge: Long = 0L
+    private var minPlayers: Int = 4
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:38]: Why are these longs

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+import at.hannibal2.skyhanni.utils.StringUtils.matchMatcher
+import at.hannibal2.skyhanni.utils.TabListData
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraft.client.Minecraft
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.math.abs
+
+class CrystalHollowsLobbyAgeWarning {
+    private val config get() = SkyHanniMod.feature.mining.crystalHollowsLobbyAgeWarning
+    private val mc get() = Minecraft.getMinecraft()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:23]: no

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+    @SubscribeEvent
+    fun onTabListUpdate(event: TabListUpdateEvent) {
+        if (!isEnabled()) return
+        processPlayerCount()
+    }
+
+    @SubscribeEvent
+    fun onIslandChange(event: IslandChangeEvent) {
+        if (event.newIsland != THE_CRYSTAL_HOLLOWS) {
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:84]: why not just reset on any island change?

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            oldLobbyAge = 0L
+        }
+    }
+    
+    @SubscribeEvent
+    fun onRepoReload(event: RepositoryReloadEvent) {
+        val data = event.getConstant<CrystalHollowsLobbyAgeWarningJson>("CrystalHollowsLobbyAgeWarning")
+        minPlayers = data.minPlayers
+        minLobbyAgeMCDays = data.minLobbyAgeMCDays
+        maxLobbyAgeMCDays = data.maxLobbyAgeMCDays
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:96]:
```suggestion
        minLobbyAge = data.minLobbyAge
        maxLobbyAge = data.maxLobbyAge
```
Dont need to specify that

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+    private fun processLobbyAge() {
+        val lobbyAge = getLobbyAgeInMinecraftDays()
+        if (oldLobbyAge == lobbyAge) return
+        oldLobbyAge = lobbyAge
+        if (config.lobbyAgeReminders && !lobbyIsPastAgeThreshold) LorenzUtils.chat("§aThis Crystal Hollows lobby is currently at Day $lobbyAge, ${StringUtils.optionalPlural(number = abs(minLobbyAgeMCDays - lobbyAge).toInt(),"day", plural = "days")} away from Day $minLobbyAgeMCDays.")
+        if (lobbyAge in minLobbyAgeMCDays..< maxLobbyAgeMCDays && !lobbyIsPastAgeThreshold) {
+            lobbyIsPastAgeThreshold = true
+            LorenzUtils.chat("This Crystal Hollows lobby has reached Day $minLobbyAgeMCDays. It no longer accepts new players, §cand will shut down on Day $maxLobbyAgeMCDays or when there are fewer than $minPlayers people left.")
+        } else if (lobbyAge >= maxLobbyAgeMCDays) LorenzUtils.chat("§cThis Crystal Hollows lobby has reached Day $maxLobbyAgeMCDays. §4§lIt will shut down very soon.")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:66]: This is unreadable also use this https://github.com/hannibal002/SkyHanni/pull/975

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt (hunk 14 above)
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:42]: also should be called getLobbyAge

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    private var minPlayers: Int = 4
+    private var minLobbyAgeMCDays: Int = 18
+    private var maxLobbyAgeMCDays: Int = 25
+
+    private fun getLobbyAgeInMinecraftDays(): Long = (mc.theWorld.worldTime) / TICKS_PER_MC_DAY //world day changes at 6AM in skyblock time
+
+    private fun processPlayerCount() {
+        for (line in TabListData.getTabList()) {
+            playerCountTabListPattern.matchMatcher(line) {
+                val playerCount = group("playerCount").formatNumber()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:47]: Can just use .toInt()

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt (hunk 13 above)
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:53]: this is also confusing and long and idk

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
changes

File: src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+                else LorenzUtils.chat("There are $playerCount players remaining in this Crystal Hollows lobby. §cIt will shut down when there are fewer than $minPlayers people left, or when this lobby reaches Day $maxLobbyAgeMCDays.")
+            }
+        }
+    }
+
+    private fun processLobbyAge() {
+        val lobbyAge = getLobbyAgeInMinecraftDays()
+        if (oldLobbyAge == lobbyAge) return
+        oldLobbyAge = lobbyAge
+        if (config.lobbyAgeReminders && !lobbyIsPastAgeThreshold) LorenzUtils.chat("§aThis Crystal Hollows lobby is currently at Day $lobbyAge, ${StringUtils.optionalPlural(number = abs(minLobbyAgeMCDays - lobbyAge).toInt(),"day", plural = "days")} away from Day $minLobbyAgeMCDays.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/mining/crystalhollows/CrystalHollowsLobbyAgeWarning.kt:62]: `StringUtils.optionalPlural` does no longer exist

**Review by hannibal002 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**